src/intersections/boxes.jl

```diff
@@ -34,3 +34,47 @@ function intersecttype(f::Function, b1::Box{Dim,T}, b2::Box{Dim,T}) where {Dim,T
     return FaceTouchingBoxes(Box(u, v)) |> f
   end
 end
+
+"""
+    intersecttype(function, ray, box)
+
+Calculate the intersection type of a `ray`` and a `box`` and apply function `f` to it.
+
+## References
+
+* Williams A, Barrus S, Morley R K, et al., 2005. [An efficient and robust ray-box
+  intersection algorithm]
+  (https://dl.acm.org/doi/abs/10.1145/1198555.1198748)
+"""
+function intersecttype(f::Function, r::Ray{Dim,T}, b::Box{Dim,T}) where {Dim,T}
+  invdir = one(T) ./ direction(r)
+  lo, up = coordinates.(extrema(b))
+  orig = coordinates(origin(r))
+
+  tmin = zero(T)
+  tmax = typemax(T)
+
+  # check for intersection with slabs along with each axis
+  for i in 1:Dim
+    imin = (lo[i] - orig[i]) * invdir[i]
+    imax = (up[i] - orig[i]) * invdir[i]
+
+    # swap if necessary
+    invdir[i] < zero(T) && ((imin, imax) = (imax, imin))
+    
+    (tmin > imax || imin > tmax) && return NoIntersection() |> f
+    
+    tmin = max(tmin, imin)
+    tmax = min(tmax, imax)
+  end
+
+  if tmin ≈ tmax
+    point = r(tmin)
+    return TouchingRayBox(point) |> f
+  elseif tmin < tmax
+    segment = Segment(r(tmin), r(tmax))
+    return CrossingRayBox(segment) |> f
+  else
```

Comment:
Hi @HaoxuanGuo can you please comment on this last branch? You had a comment that said "May got NaN", why is that? Also what happens when the `invdir` explodes to `Inf` because one of the coordinates is zero?

Reply:
Found this comment in the cited paper:

![image](https://user-images.githubusercontent.com/3345261/143774124-15073e1e-29f4-4472-a783-80e93eee292e.png)

Reply:
BTW very interesting to find out a paper specifically for this intersection type.

Reply:
@juliohm This is a very special situation - the ray is on a face of the box. In fact, it is just no intersection but with a special branch condition.

For example, this code will hit the last branch.

```julia
using Meshes
r = Ray((1.5,0,0), (-1,1,0.0))
b = Box((0,0,0), (1,1,1))
b ∩ r
```

Reply:
Isn't this case an intersection? I mean strictly speaking the set of points is non empty in this case and it forms a new ray, right?